ec2/utils.py

```diff
@@ -76,120 +215,267 @@ def get_instance_state(instance, ctx):
+    kplist = get_target_aws_resource_id(relationship_type, ctx=ctx)
+    if kplist:
+        return kplist[0]
+    else:
+        return kplist
 
 
-def get_private_dns_name(instance, timeout_length):
-    """ returns the private_dns_name variable for a given instance
-    """
-    return get_instance_attribute(instance,
-                                  'private_dns_name', timeout_length)
+def get_attached_security_group_ids(ctx):
+    relationship_type = 'instance_connected_to_security_group'
 
+    return get_target_aws_resource_id(relationship_type, ctx=ctx)
 
-def get_public_dns_name(instance, timeout_length):
-    """ returns the public_dns_name variable for a given instance
+
+def get_target_aws_resource_id(relationship_type, ctx):
+    """ This loops through the relationships of type and returns
+        targets of those relationships.
     """
-    return get_instance_attribute(instance,
-                                  'public_dns_name', timeout_length)
 
+    ids = []
 
-def get_private_ip_address(instance, timeout_length):
-    """ returns the private_ip_address variable for a given instance
-    """
-    return get_instance_attribute(instance,
-                                  'private_ip_address', timeout_length)
+    if not hasattr(ctx.instance, 'relationships'):
+        ctx.logger.info('Skipping attaching relationships, '
+                        'because none are attached.')
+        return ids
 
+    for r in ctx.instance.relationships:
+        if relationship_type in r.type:
+            ids.append(
+                r.target.instance.runtime_properties['aws_resource_id'])
+
+    return ids
 
-def get_public_ip_address(instance, timeout_length):
-    """ returns the public_ip_address variable for a given instance
-    """
-    return get_instance_attribute(instance,
-                                  'public_ip_address', timeout_length)
 
+def get_security_group_from_id(group_id, ctx):
+    ec2_client = connection.EC2ConnectionClient().client()
+    ctx.logger.debug('Getting Security Group by ID: {0}'.format(group_id))
+
+    try:
+        group = ec2_client.get_all_security_groups(group_ids=group_id)
+    except boto.exception.EC2ResponseError as e:
+        group = get_security_group_from_name(group_id, ctx=ctx)
+        raise NonRecoverableError('Error. '
+                                  'Failed to group by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+    except boto.exception.BotoServerError as e:
+        raise NonRecoverableError('Error. '
+                                  'Failed to group by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
 
-def validate_group(group, ctx):
-    ctx.logger.debug('Testing if group with identifier '
-                     ' {0} exists in this account.'.format(group))
-    groups = get_security_group_from_id(group, ctx)
+    if len(group) < 1:
```

Comment:
I started doing what you suggested about keypair (return group[0] if group else group ) for everything